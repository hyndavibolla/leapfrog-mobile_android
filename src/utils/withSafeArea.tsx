import React, { useMemo } from 'react';
import { StyleSheet, StatusBar, View, SafeAreaView as SafeView, StatusBarStyle } from 'react-native';
import { SafeAreaView, Edge } from 'react-native-safe-area-context';

import { useTestingHelper } from '_utils/useTestingHelper';
import { COLOR } from '_constants';

const styles = StyleSheet.create({
  safeArea: {
    flex: 1,
    backgroundColor: COLOR.WHITE
  },
  container: {
    flex: 1
  },
  statusBar: {
    height: StatusBar.currentHeight
  }
});

type ConfigHeader = {
  onlyBlueStatusBar?: boolean;
  isFixedHeader?: boolean;
  color?: COLOR;
  barStyle?: StatusBarStyle;
};

export const withSafeArea =
  <P extends object>(
    Component: React.ComponentType<P>,
    withHeader: boolean,
    configHeader?: ConfigHeader,
    edges: ReadonlyArray<Edge> = ['top', 'left', 'right', 'bottom']
  ) =>
  (props: P) => {
    configHeader = { onlyBlueStatusBar: false, color: COLOR.PRIMARY_BLUE, ...configHeader };
    const { getTestIdProps } = useTestingHelper('with-safe-area');

    const appStatusBar = useMemo(
      () =>
        configHeader.onlyBlueStatusBar ? (
          <View style={[styles.statusBar, { backgroundColor: configHeader.color }]} {...getTestIdProps('statusbar-blue')}>
            <SafeView>
              <StatusBar translucent backgroundColor={configHeader.color} barStyle={configHeader.barStyle ?? 'light-content'} />
            </SafeView>
          </View>
        ) : (
          <StatusBar translucent backgroundColor={COLOR.TRANSPARENT} barStyle={configHeader.barStyle ?? 'dark-content'} />
        ),
      [getTestIdProps]
    );

    return configHeader.onlyBlueStatusBar || configHeader.isFixedHeader ? (
      <View style={styles.container}>
        {appStatusBar}
        <View style={styles.container}>
          <Component {...(props as P)} />
        </View>
      </View>
    ) : (
      <>
        {appStatusBar}
        <SafeAreaView edges={withHeader ? ['left', 'right', 'bottom'] : edges} style={styles.safeArea} {...getTestIdProps('statusbar-default')}>
          <Component {...(props as P)} />
        </SafeAreaView>
      </>
    );
  };
