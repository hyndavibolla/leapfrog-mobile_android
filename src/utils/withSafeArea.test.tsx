import React from 'react';
import { withSafeArea } from './withSafeArea';
import { render } from '@testing-library/react-native';

describe('withSafeArea', () => {
  it('should return  component wrapped in a SafeAreaView with header support', () => {
    const Component = withSafeArea(() => <div>this is the wrapped component's body</div>, true);
    const { toJSON } = render(<Component />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('should return component wrapped in a SafeAreaView without header support', () => {
    const Component = withSafeArea(() => <div>this is the wrapped component's body</div>, false);
    const { getByTestId } = render(<Component />);
    expect(getByTestId('with-safe-area-statusbar-default')).toBeTruthy();
  });

  it('should return component wrapped without header and with statusbar blue', () => {
    const Component = withSafeArea(() => <div>this is the wrapped component's body</div>, false, { onlyBlueStatusBar: true });
    const { getByTestId } = render(<Component />);
    expect(getByTestId('with-safe-area-statusbar-blue')).toBeTruthy();
  });

  it('should return component wrapped in a SafeAreaView with custom edge', () => {
    const Component = withSafeArea(() => <div>this is the wrapped component's body</div>, false, { onlyBlueStatusBar: false }, ['left', 'right']);
    const { toJSON } = render(<Component />);
    expect(toJSON()).toMatchSnapshot();
  });
});
