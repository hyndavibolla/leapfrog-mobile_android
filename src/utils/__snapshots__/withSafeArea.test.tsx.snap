// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withSafeArea should return  component wrapped in a SafeAreaView with header support 1`] = `
<RNCSafeAreaView
  accessibilityLabel="with-safe-area-statusbar-default"
  accessible={true}
  edges={
    Array [
      "left",
      "right",
      "bottom",
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
  testID="with-safe-area-statusbar-default"
>
  <div>
    this is the wrapped component's body
  </div>
</RNCSafeAreaView>
`;

exports[`withSafeArea should return component wrapped in a SafeAreaView with custom edge 1`] = `
<RNCSafeAreaView
  accessibilityLabel="with-safe-area-statusbar-default"
  accessible={true}
  edges={
    Array [
      "left",
      "right",
    ]
  }
  style={
    Object {
      "backgroundColor": "#FFFFFF",
      "flex": 1,
    }
  }
  testID="with-safe-area-statusbar-default"
>
  <div>
    this is the wrapped component's body
  </div>
</RNCSafeAreaView>
`;
